packages/react-native/ReactAndroid/src/main/java/com/facebook/react/LazyReactPackage.kt

```diff
@@ -0,0 +1,140 @@
+import com.facebook.react.module.model.ReactModuleInfo
+import com.facebook.react.module.model.ReactModuleInfoProvider
+import com.facebook.react.uimanager.ViewManager
+import com.facebook.systrace.Systrace.TRACE_TAG_REACT
+import com.facebook.systrace.SystraceMessage
+
+/** React package supporting lazy creation of native modules. */
+@Deprecated("This class is deprecated, please use BaseReactPackage instead.")
+@LegacyArchitecture
+public abstract class LazyReactPackage public constructor() : ReactPackage {
+    /**
+     * We return an iterable
+     *
+     * @param reactContext context
+     * @return [<] that contains all native modules registered for the context
+     */
+    public fun getNativeModuleIterator(
+        reactContext: ReactApplicationContext?
+    ): Iterable<ModuleHolder> {
+        val reactModuleInfoMap: Map<String, ReactModuleInfo> =
+            reactModuleInfoProvider.getReactModuleInfos()
+        val nativeModules = getNativeModules(reactContext)
+
+        return object : Iterable<ModuleHolder> {
+            override fun iterator(): Iterator<ModuleHolder> {
+                var position = 0
+
+                return object : Iterator<ModuleHolder> {
+                    override fun hasNext(): Boolean {
+                        return position < nativeModules.size
+                    }
+
+                    override fun next(): ModuleHolder {
+                        val moduleSpec = nativeModules[position++]
+                        val name = moduleSpec.name
+
+                        return reactModuleInfoMap[name].let moduleHolder@{ reactModuleInfo ->
+                            if (reactModuleInfo == null) {
+                                val module: NativeModule
+                                ReactMarker.logMarker(
+                                    ReactMarkerConstants.CREATE_MODULE_START,
+                                    name,
+                                )
+                                try {
+                                    module = moduleSpec.provider.get()
+                                } finally {
+                                    ReactMarker.logMarker(ReactMarkerConstants.CREATE_MODULE_END)
+                                }
+                                return@moduleHolder ModuleHolder(module)
+                            } else {
+                                return@moduleHolder ModuleHolder(
+                                    reactModuleInfo,
+                                    moduleSpec.provider,
+                                )
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * @param reactContext react application context that can be used to create modules
+     * @return list of module specs that can create the native modules
+     */
+    public abstract fun getNativeModules(reactContext: ReactApplicationContext?): List<ModuleSpec>
```

Comment:
Let's try to keep the same nullability for the arguments as in the original implementation

```suggestion
    public abstract fun getNativeModules(reactContext: ReactApplicationContext): List<ModuleSpec>
```